### PR TITLE
Implementing project-level and namespace-level quota cap

### PR DESCRIPTION
**Pull request by bnsblue**

### CODE REVIEW

File: pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go [hunk 1]
```diff
@@ -184,16 +199,45 @@ func (r *RedisResourceManager) GetID() string {
 	return RedisSetKeyPrefix
 }
 
+func (r *RedisResourceManager) getNamespaceAllocatedCount(ctx context.Context, client RedisClient, namespace pluginCore.ResourceNamespace, allocationToken string) (int, error) {
+	// TODO: redis's SMembers potentially leads to performance degradation at Redis side when there are too many members in the set.
+	// 		 If that happens, we should replace SMembers with Sscan, which needs extra work to deal with duplicated members
+	tokenPrefix, err := extractTokenPrefix(allocationToken)
```
[the remark this answers is not in the document]
Comment (bnsblue) [reply pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go]: I don't think what you described can achieve the functionality we want to have. What we want is counting the number of strings with a specific prefix -- we cannot "build the final allocation token" in this case.

File: pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go [hunk 2]
```diff
@@ -184,16 +175,52 @@ func (r *RedisResourceManager) GetID() string {
 	return RedisSetKeyPrefix
 }
 
-func (r *RedisResourceManager) AllocateResource(ctx context.Context, namespace pluginCore.ResourceNamespace, allocationToken string) (
-	pluginCore.AllocationStatus, error) {
+func (r *RedisResourceManager) checkAgainstOneConstraint(_ context.Context, allAllocated []string,
+	constraint ComposedResourceConstraint) bool {
+	var count int64 = 0
+	for _, allocated := range allAllocated {
+		if does := strings.HasPrefix(allocated, constraint.TargetedPrefixString); does {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go]:
```suggestion
		if strings.HasPrefix(allocated, constraint.TargetedPrefixString) {
```

File: pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go [hunk 3]
```diff
@@ -184,16 +175,52 @@ func (r *RedisResourceManager) GetID() string {
...
+			count++
+		}
+		if count >= constraint.Value {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go]: Nit: move this line/check as a function on the constraint type.

File: pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go [hunk 4]
```diff
@@ -184,16 +175,52 @@ func (r *RedisResourceManager) GetID() string {
...
+			return false
+		}
+	}
+	return true
+}
+
+func (r *RedisResourceManager) checkAgainstConstraints(ctx context.Context, client RedisClient, resource pluginCore.ResourceNamespace,
+	constraints []ComposedResourceConstraint) (bool, int, error) {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go]:
```suggestion
	constraints []ComposedResourceConstraint) (allowed bool, violatedConstraintIndex int, err error) {
```

File: pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go [hunk 5]
```diff
@@ -203,19 +230,36 @@ func (r *RedisResourceManager) AllocateResource(ctx context.Context, namespace p
...
 
-	size, err := r.client.SCard(string(namespace)).Result()
+	size, err := r.client.SCard(string(namespace))
 	if err != nil {
 		logger.Errorf(ctx, "Error getting size of Redis set %v", err)
 		return pluginCore.AllocationUndefined, err
 	}
 
-	if size > int64(namespacedResource.quota) {
+	if size >= int64(namespacedResource.quota) {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go]: nit: should this be represented as a "constraint" as well and use the globalConstraint.IsAllowed(size) check for consistency?

File: pkg/controller/nodes/task/resourcemanager/resourceconstraints.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package resourcemanager
+
+import (
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
+	pluginCore "github.com/lyft/flyteplugins/go/tasks/pluginmachinery/core"
+)
+
+type ComposedResourceConstraint struct {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/resourcemanager/resourceconstraints.go]: nit: docs
also, what does the 'composed' mean in the name?
aslo, can you rename the file to match the type name?

File: pkg/controller/nodes/task/resourcemanager/resourcemanager.go [hunk 7]
```diff
@@ -12,23 +12,36 @@ import (
...
+// appending the actual token string to the prefix using '-' separator
+func (t TokenPrefix) append(tokenWithoutPrefix string) Token {
+	return Token(fmt.Sprintf("%s%s%s", t, tokenNamespaceSeparator, tokenWithoutPrefix))
 }
 
-func composeExecutionUrn(id *core.TaskExecutionIdentifier) string {
-	return execUrnPrefix + execUrnSeparator + id.GetNodeExecutionId().GetExecutionId().GetProject() +
-		execUrnSeparator + id.GetNodeExecutionId().GetExecutionId().GetDomain() + execUrnSeparator + id.GetNodeExecutionId().GetExecutionId().GetName()
+// extending the prefix using ':' separator
+func (t TokenPrefix) extend(prefixPart string) TokenPrefix {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/resourcemanager/resourcemanager.go]: When do you use extend vs append? the docs are not clear to me of the implications...

File: pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go (hunk 5 above)
Comment (bnsblue) [reply pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go]: refactored

File: pkg/controller/nodes/task/resourcemanager/resourcemanager.go (hunk 7 above)
Comment (bnsblue) [reply pkg/controller/nodes/task/resourcemanager/resourcemanager.go]: Refactored. Now I have TokenPrefix.extend() and Token.prepend(). Is it clearer?

File: pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go (hunk 3 above)
Comment (bnsblue) [reply pkg/controller/nodes/task/resourcemanager/redis_resourcemanager.go]: Refactored

**Pull request merged**